### ISSUE TITLE
Allow types other than StdoutLock so that a tty can be used in place of Stdout

### ISSUE THREAD

**Issue by norcalli** [ftilde/unsegen#1]
If one wants to use stdin for a data pipe, then we must allocate using `/dev/tty`, or, in this case, `termion::get_tty()`, which returns a `File`. In this case, I believe that `stdout` will be rendered unusable by `Terminal` because I get the error `"Inappropriate ioctl for device"`, which means that by reading `stdin`, `stdout` is no longer a tty (I'm not too familiar with tty, to be honest). I tried bypassing this by `dup2`ing the File's fd onto stdout and using stdoutlock anyway, but that isn't working either.

**Comment by ftilde** [comment]
Hmm, I'm not entirely sure I understand why what exactly is going wrong, yet. I'm not really familiar with all details of ttys either, unfortunately.

Could you maybe post what you've got so far as a small example so I can play around with it?

It definitely sounds useful to be able to still read from stdin!

**Comment by ftilde** [comment]
So, I've dug a bit deeper into this even without an example. On master I modified `Terminal` to be used with any `Write`, not just `StdoutLock`. Now we can use `/dev/tty` directly with `Terminal`:

```
fn main() {
    let tty_w = get_tty().unwrap();
    let mut term = Terminal::new(tty_w);
    // ...
}
```

However, even here the terminal setup fails if we pipe to stdin, again with `"Inappropriate ioctl for device"`. This happens when we try to switch the terminal to raw mode using termion. Still not sure why that fails, though.

**Comment by norcalli** [comment]
Sorry I've been busy, I was planning to make a followup, but your example is pretty much what I wanted to do. The example use case is something like [jsonfui](https://github.com/AdrianSchneider/jsonfui), which displays json data piped in from stdin. In fact, my exact use case was building a more responsive version of jsonfui, since after a few hundred nodes it starts to really slow down, and I deal with JSON data in the megabytes to gigabytes range.

I'll give it another shot and try implementing a basic UI with just termion to better understand tty's. I'm assuming this is a common problem.

**Comment by norcalli** [comment]
@ftilde I figured out why this was happening, btw. It's because `termion` has a bug. The `into_raw_mode()` behind the scenes was hard coding to only work on fd 0/1. Therefore, the `get_tty` new tty was receiving an incorrect ioctl since it wasn't a tty. I wrote a library to fix this that I'll be publishing soon.

**Comment by ftilde** [comment]
Thanks for reporting back! After your message I took a look at the termion repo again and found this [merge request](https://gitlab.redox-os.org/redox-os/termion/merge_requests/124) from 2017 which addresses the same problem. Not sure why it isn't merged yet. Possibly because of the API change.

Looking forward to hear from your library, though. Maybe we can use this to work around the problem in unsegen.

**Comment by norcalli** [comment]
@ftilde https://crates.io/crates/raw_tty let me know if the API makes sense/helps.

**Comment by ftilde** [comment]
@norcalli Thank you! I've switched to using raw_tty for setting up the terminal in unsegen now.
Regarding the API: Does not seem to be lacking anything for, but I'm gonna an issue shortly.

**Comment by ftilde** [comment]
Oh, and sidenote: In a way this issue would also be closed by updating termion to 1.5.3 because of this [MR](https://gitlab.redox-os.org/redox-os/termion/merge_requests/161). However, as mentioned there, this may not work if you (for some reason) create a `Terminal` with something else then stdout, so I'm sticking with raw_tty for now.
